tools/astarte_e2e/lib/astarte_e2e/client.ex

```diff
@@ -0,0 +1,406 @@
+    Logger.info("Neglecting msg. Event: #{inspect(event)}, payload: #{inspect(payload)}.")
+
+    {:ok, state}
+  end
+
+  def handle_reply(_topic, event, payload, _transport, state) do
+    Logger.info("Handling reply. Event: #{inspect(event)}, payload: #{inspect(payload)}.")
+
+    {:ok, state}
+  end
+
+  def handle_join_error(topic, _payload, _transport, state) do
+    Logger.warn("Stopping process.", tag: "astarte_e2e_join_error")
+    {:stop, {:error, {:join_failed, topic}}, state}
+  end
+
+  def handle_info(:timeout, _transport, state) do
+    :telemetry.execute(
+      [:astarte_end_to_end, :astarte_platform, :status],
+      %{health: 0}
+    )
+
+    Logger.error("Request timed out.", tag: "astarte_e2e_request_timeout")
+    {:stop, :timeout, state}
+  end
+
+  def handle_info({:watch_error, reason}, _transport, _state) do
+    {:error, reason}
+  end
+
+  def handle_info(:try_connect, _transport, state) do
+    if state.connection_attempts > 0 do
+      updated_attempts = state.connection_attempts - 1
+      updated_state = %{state | connection_attempts: updated_attempts}
+      {:connect, updated_state}
+    else
+      Logger.warn(
+        "Cannot establish a connection after #{inspect(@connection_attempts)} attempts. Closing application.",
+        tag: "astarte_e2e_client_connection_failed"
+      )
+
+      System.stop(1)
+      {:stop, :connection_failed, state}
+    end
+  end
+
+  def handle_call(
+        {:verify_payload, _interface_name, _path, _value, _timestamp},
+        _from,
+        _transport,
+        %{connected: false} = state
+      ) do
+    :telemetry.execute([:astarte_end_to_end, :messages, :failed], %{})
+
+    {:reply, {:error, :not_connected}, state}
+  end
+
+  def handle_call(
+        {:verify_payload, interface_name, path, value, timestamp},
+        from,
+        _transport,
+        state
+      ) do
+    pending_messages = Map.get(state, :pending_messages)
+    pending_requests = Map.get(state, :pending_requests)
+
+    if Map.has_key?(pending_messages, {interface_name, path, value}) do
+      {reception_timestamp, new_pending_messages} =
+        Map.pop(pending_messages, {interface_name, path, value})
+
+      dt_ms = reception_timestamp - timestamp
+      new_state = Map.put(state, :pending_messages, new_pending_messages)
+
+      :telemetry.execute(
+        [:astarte_end_to_end, :messages, :round_trip_time],
+        %{duration_seconds: dt_ms / 1_000}
+      )
+
+      :telemetry.execute(
+        [:astarte_end_to_end, :astarte_platform, :status],
+        %{health: 1}
+      )
+
+      Logger.info("Round trip time = #{inspect(dt_ms)} ms.")
+
+      {:reply, {:ok, {:round_trip_time_ms, dt_ms}}, new_state}
+    else
+      new_pending_requests =
+        Map.put(pending_requests, {interface_name, path, value}, {timestamp, from})
+
+      new_state = Map.put(state, :pending_requests, new_pending_requests)
+
+      {:noreply, new_state}
```

Comment:
`verify_device_payload` is called with `:infinity` timeout, so if the correspondig message never arrives the call will never returned and it will get stuck forever.

You put the `handle_info` for the timeout in place, but you probably forgot to add
```elixir
{:noreply, new_state, timeout}
```
here.

Then again, this would mean that if _one_ request times out, the scheduler is going to crash. Do we want to be so dramatic in handling a single timeout? I would probably handle it by returning `{:error, :timeout}` to the caller.